XCH4 ESA CMUG diagnostics (subset of the MPQB diagnostics)

hb326 opened the pull request:
First diagnostic to use ESA CMUG XCH4 data for evaluating CMIP datasets.
Diagnostics are a subset of the C3S MPQB diagnostics, some with minor adjustments.


* * *

**Checklist for technical review**

- [ ] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
- [x] The pull request has a descriptive title that can be used as a one line summary in the [changelog](https://docs.esmvaltool.org/en/latest/changelog.html)
- [x] The code is composed of functions of no more than 50 lines and uses meaningful names for variables and follows the [style guide](https://docs.esmvaltool.org/en/latest/community/introduction.html#code-style)
- [x] [Documentation](https://docs.esmvaltool.org/en/latest/community/introduction.html#documentation) is available
- [x] Please use `yamllint` to check that your YAML files do not contain mistakes
- [ ] (Only if really necessary) Add any additional dependencies needed for the diagnostic script to setup.py, esmvaltool/install/R/r_requirements.txt or esmvaltool/install/Julia/Project.toml (depending on the language of your script) and also to package/meta.yaml for conda dependencies (includes Python and others, but not R/Julia). Also check that the license of the dependency you want to add and any of its dependencies are compatible with [Apache2.0](https://github.com/ESMValGroup/ESMValTool/blob/master/LICENSE).

New or updated [recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html):

- [x] Documentation for the recipe/diagnostic is available in the `doc/sphinx/source/recipes` folder and an entry has been added to `index.rst`
- [x] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added and no warnings related to provenance are generated when running the recipe

Automated checks pass, status can be seen below the pull request:

- [x] Circle/CI tests pass. If the tests are failing, click the `Details` link to find out why.
- [ ] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
- [x] The documentation is building successfully on readthedocs and looks well formatted, click the `Details` link to see it.


**Checklist for scientific review**

New or updated [recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html):

- [x] [The documentation](https://docs.esmvaltool.org/en/latest/recipes/index.html) for the new/updated recipes/diagnostics clearly describes what the recipe does and how to use it
- [x] The recipe runs successfully on your own machine/cluster or with the [`@esmvalbot`](https://github.com/apps/esmvalbot) without any modifications to the recipe and with all data specified in the recipe
- [x] The figure(s) and data look as expected from the literature and/or the paper that is reproduced by the recipe
- [ ] The code contains comments with references to formulas, figures, tables, etc. that are used from papers/online resources

* * *

If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.

This pull request needs ESMValCore pull request 1428 (https://github.com/ESMValGroup/ESMValCore/pull/1428) to be merged to be able to run the recipe.

valeriupredoi commented:
@hb326 would you have time to finalize this in time for the code freeze?

hb326 commented:
> 
> 
> @hb326 would you have time to finalize this in time for the code freeze?

@valeriupredoi: no, that will not be finished by the time for the code freeze. Hopefully it will be good to go for the next release.

valeriupredoi commented:
@hb326 @bascrezee I'd be happy to start reviewing this so it materializes into an active-ready PR, just lemme know when I should start looking at the code pls :beer:

hb326 commented:
Still have to figure out some of the failing checks.

valeriupredoi commented:
> Still have to figure out some of the failing checks.

@hb326 quite a few of them will have been fixed by merging the latest `main` here :+1:

hb326 commented:
> > Still have to figure out some of the failing checks.
> 
> @hb326 quite a few of them will have been fixed by merging the latest `main` here 👍

I keep forgetting to do that! Will do it now and see what will happen. Ta!

hb326 commented:
> > > Still have to figure out some of the failing checks.
> > 
> > 
> > @hb326 quite a few of them will have been fixed by merging the latest `main` here 👍
> 
> I keep forgetting to do that! Will do it now and see what will happen. Ta!

It seems like I had indeed merged the latest main into the branch. At least when I try to merge main into the branch now, it tells me that it is up to date...

valeriupredoi commented:
the remote commit history is not showing it, did you push? Also make sure you grab the latest `main` from remote via ` git pull origin main` when you are on checked in on `main`, Birgit :smiley:

hb326 commented:
> the remote commit history is not showing it, did you push? Also make sure you grab the latest `main` from remote via ` git pull origin main` when you are on checked in on `main`, Birgit 😃

Ha! I did merge v2.3 of "main" in this branch here, but not v2.4. Needed to update that. Which I did, and voila, merging is happening.  :)

valeriupredoi commented:
good work Birgit! Now only that lauer_adj recipe needs a title, and these are Flake8 errors:
```
FAILED esmvaltool/diag_scripts/mpqb/mpqb_lineplot_anncyc.py::FLAKE8
FAILED esmvaltool/diag_scripts/mpqb/mpqb_utils.py::FLAKE8
FAILED esmvaltool/diag_scripts/mpqb/mpqb_lineplot.py::FLAKE8
FAILED esmvaltool/diag_scripts/mpqb/mpqb_lineplot_growthrate.py::FLAKE8
FAILED esmvaltool/diag_scripts/shared/_validation.py::FLAKE8
```

Note that the flake8 issues are masked (see https://github.com/ESMValGroup/ESMValCore/issues/1405 ) but if you run an `isort` first to make sure the imports are all in the correct order, then run a `prospector` on each of those scripts you'll be able to find out why flake8 is complaining :beer:

axel-lauer commented:
The documentation looks good, but maybe also add some information on the observational dataset (section "Observations and reformat scripts")? Now running the recipe and then looking at the results...

axel-lauer commented:
I just tested this new recipe. The recipe runs fine and produces a number of plots that look good (yay!). Here are a few small things I noticed:

- The observational dataset is currently commented out in the recipe. Since this is an ESA CMUG recipe, I would recommend to include the reference dataset by default.
- The references currently only include Righi et al. (2020). Is there possibly anything else to be cited for the diagnostics?
- Independant of the region processed (global, NH, SH), the provenance record always gives "domains('global')".
- The CMUG project could be added as project to the provenance record.
- The caption in the provenance record says "Global" (independant of the actual region).
- The caption in the provenance records for the growth rates says "Global mean time series" --> maybe change into something like "global mean annual growth rates"?
- The attribute "statistics" in the provenance record for annual growth rates gives "mean". Is that intended?
- Would be nice to add some information on the observational dataset and CMORIzer script (section "Observations and reformat scripts") to the doumentation.

hb326 commented:
> The documentation looks good, but maybe also add some information on the observational dataset (section "Observations and reformat scripts")? Now running the recipe and then looking at the results...

Just added this information to the documentation.

hb326 commented:
> 

- I have commented in the observational dataset.
- I cannot think of another reference for these diagnostics. They are very basic in the sense that they are just lineplots without too much statistics behind it...
- I adjusted the provenance, so that it states next to "global" also "nh" and "sh"
- It looks to me like "cmug" is already added as project in the provenance. At least it is listed in the provenance record where it is created. Hmm, there might be something going wrong...
- I changed the caption in the provenance records for the growth to "global mean annual growth rates" as suggested
- I adjusted the "statistics" for the growth rates plot
- I added a section to the documentation about the observations and the derivation of the variable

axel-lauer commented:
@hb326 Thanks for working hard on this! I think most of the point I raised earlier are fine now. What I still fine a bit strange is the following in the provenance:

* the attribute 'domains' now always says 'global, northern hemisphere, southern hemisphere' instead of just one of the latter
* the attribute 'caption' continues to say "global ..." instead of giving the specific region

Other than that, I think all looks good! Nice work!

axel-lauer commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot.py:
```diff
@@ -0,0 +1,148 @@
+    fig.savefig(filename, bbox_inches='tight')
+
+    caption = (
+        f"Global mean time series of {dataset_cfg['variable_group']} between "
```
```suggestion
        f"Time series of the domain average of {dataset_cfg['variable_group']} between "
```

axel-lauer commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot.py:
```diff
@@ -0,0 +1,148 @@
+    record = {
+        'caption': caption,
+        'statistics': ['mean'],
+        'domains': ['global', 'nh', 'sh'],
```
```suggestion
        'domains': ['user'],
```

axel-lauer reviewed (approved):
Receipe runs fine, documentation and output look good!

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_cfg_xch4.yml:
```diff
@@ -0,0 +1,41 @@
+#
+# Configuration related to plotting of different ECVs
+# for usage in the different MPQB diagnostics.
+#
+# mpqb_cfg.yml
```
```suggestion
```

schlunma commented on doc/sphinx/source/recipes/recipe_mpqb_xch4.rst:
```diff
@@ -0,0 +1,121 @@
+   * ``mpqb_cfg_xch4.yml``: In this file additional information for the used datasets are defined and stored, e.g. alias of the dataset name and the color that is used to display the dataset in the figures
+   * ``mpqb_utils.yml``: In this file the preparations for the dataset displays are made.
+
+#. Script <mpqb_lineplot.py>
```
```suggestion
#. Script ``mpqb_lineplot.py``
```

schlunma commented on doc/sphinx/source/recipes/recipe_mpqb_xch4.rst:
```diff
@@ -0,0 +1,121 @@
+   
+   * no settings for the variable required
+
+#. Script <mpqb_lineplot_anncyc.py>
```
```suggestion
#. Script ``mpqb_lineplot_anncyc.py``
```

schlunma commented on doc/sphinx/source/recipes/recipe_mpqb_xch4.rst:
```diff
@@ -0,0 +1,121 @@
+   
+   * no settings for the variable required
+
+#. Script <mpqb_lineplot_growthrate.py>
```
```suggestion
#. Script ``mpqb_lineplot_growthrate.py``
```

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot.py:
```diff
@@ -0,0 +1,149 @@
+
+def main(cfg):
+    """Create lineplot."""
+    ylims = [cfg.pop('y0', None), cfg.pop('y1', None)]
```
It looks like `y0` and `y1` are possible optional options for this diagnostic which are not listed in the documentation

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot_anncyc.py:
```diff
@@ -0,0 +1,116 @@
+
+def main(cfg):
+    """Create lineplot."""
+    ylims = [cfg.pop('y0', None), cfg.pop('y1', None)]
```
Same comment here: It looks like `y0` and `y1` are possible optional options for this diagnostic which are not listed in the documentation

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot.py:
```diff
@@ -0,0 +1,149 @@
+        label = get_mpqb_cfg('datasetname', alias)
+        color = get_mpqb_cfg('datasetcolor', alias)
+
+        # iris.quickplot.plot(cube, label=label, color=color,
+        #                     linestyle='dotted')
+        # iris.quickplot.plot(cube, label=label, color=color,
+        #                     linestyle='dashed')
```
```suggestion
```

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot_anncyc.py:
```diff
@@ -0,0 +1,116 @@
+        label = get_mpqb_cfg('datasetname', alias)
+        color = get_mpqb_cfg('datasetcolor', alias)
+
+        # iris.quickplot.plot(cube, label=label, color=color,
+        #                     linestyle='dotted')
+        # iris.quickplot.plot(cube, label=label, color=color,
+        #                     linestyle='dashed')
```
```suggestion
```

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot_growthrate.py:
```diff
@@ -0,0 +1,151 @@
+        label = get_mpqb_cfg('datasetname', alias)
+        color = get_mpqb_cfg('datasetcolor', alias)
+
+        # iris.quickplot.plot(cube, label=label, color=color,
+        #                     linestyle='dotted')
+        # iris.quickplot.plot(cube, label=label, color=color,
+        #                     linestyle='dashed')
```
```suggestion
```

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_utils.py:
```diff
@@ -0,0 +1,32 @@
+def get_mpqb_cfg(cfgtype, cfgkey):
+    """cfgtype: colormap, datasetcolor, datasetname"""
+    # First read cfg file
+    cfg_filename = os.path.join(os.path.split(__file__)[0],
+                                'mpqb_cfg_xch4.yml')
```
I think it's better to rename `mpqb_cfg_xch4.yml` -> `mpqb_cfg.yml` and also rename the corresponding file. This part of the diagnostic should not as generic as possible.

schlunma commented on esmvaltool/recipes/mpqb/recipe_mpqb_xch4.yml:
```diff
@@ -0,0 +1,307 @@
+  - {dataset: CDS-XCH4, project: OBS, type: sat, version: L3, tier: 3,
+     derive: false, alias: CDS-XCH4}
+### CMIP6 (esm-hist)
+  # - {dataset: BCC-CSM2-MR, project: CMIP6, ensemble: r1i1p1f1, grid: gn,
+     # exp: esm-hist, derive: true, alias: BCC-CSM2-MR}
+  # - {dataset: CNRM-ESM2-1, project: CMIP6, ensemble: r1i1p1f2, grid: gr,
+     # exp: esm-hist, derive: true, alias: CNRM-ESM2-1}
+  # - {dataset: UKESM1-0-LL, project: CMIP6, ensemble: r1i1p1f2, grid: gn,
+     # exp: esm-hist, derive: true, alias: UKESM1-0-LL}
```
Does the recipe work when you uncomment these datasets? If yes, please uncomment; if not, might be better to remove them.

hb326 replied:
Yes, the recipe works also with these datasets. They are just commented out because you either look at the "historical" runs or the "esm-hist" runs, but normally not at both at the same time. However, since we decided to not look at these datasets in detail, I will delete these datasets in the recipe.

schlunma commented on setup.py:
```diff
@@ -52,6 +52,7 @@
         'ruamel.yaml',
         'scikit-image',
         'scikit-learn',
+        'SkillMetrics',
```
```suggestion
```

I don't find a reference to this package in any of the new diagnostics, so this can probably be removed.

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot.py:
```diff
@@ -0,0 +1,149 @@
+    ax1.xaxis.set_major_formatter(years_fmt)
+    ax1.grid(True, which='major', axis='x')
+    ax1.set_ylim(ylims)
+    ax1.set_ylabel('XCH4 (ppbv)')
+    ax1.set_title('Time series of monthly mean XCH4')
```
```suggestion
    ax1.set_ylabel(f"{cube.var_name.upper()} ({cube.units})")
    ax1.set_title(f"Time series of monthly mean {cube.var_name.upper()}")
```

It might be good to keep this diagnostic generic and not hardcode `XCH4` here. It could either be added as `cfg` option or from the cube's `var_name` and `units`.

hb326 replied:
I like the suggesttion, thanks!

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot_anncyc.py:
```diff
@@ -0,0 +1,116 @@
+    ax1.grid(True, which='major', axis='x')
+
+    ax1.set_ylim(ylims)
+    ax1.set_ylabel('XCH4 (ppbv)')
+    ax1.set_title('XCH4 annual cycle')
```
```suggestion
    ax1.set_ylabel(f"{cube.var_name.upper()} ({cube.units})")
    ax1.set_title(f"{cube.var_name.upper()} annual cycle")
```

hb326 replied:
Same here, I like the changes, thanks!

schlunma commented on esmvaltool/diag_scripts/mpqb/mpqb_lineplot_growthrate.py:
```diff
@@ -0,0 +1,151 @@
+    ax1.grid(True, which='major', axis='x')
+
+    ax1.set_ylim(ylims)
+    ax1.set_ylabel('Change XCH4 (ppbv)')
+    ax1.set_title('Time series of annual XCH4 growth rate')
```
```suggestion
    ax1.set_ylabel(f"Change in {cube.var_name.upper()} ({cube.units})")
    ax1.set_title(f"Time series of annual {cube.var_name.upper()} growth rate")
```

hb326 replied:
Also here, thanks for the suggestions, they are great.

schlunma reviewed (commented):
Cool stuff! :tada: 

I only have minor comments regarding the documentation, commented lines and specific plot labels.

axel-lauer commented:
@schlunma Thanks for your review! @hb326 I would recommend to accept all of @schlunma 's proposed changes. Then the PR can be merged. Yay!

hb326 commented:
You were a little quick, @axel-lauer, with the merging.  :)
I wanted to update the recipe figure because it is without the observations right now. Will do that in a new pull request then...

axel-lauer commented:
@hb326 ups... Sorry about this. I created a pull request with an updated figure including the ESACCI data: https://github.com/ESMValGroup/ESMValTool/pull/2502
Please take a look and approve PR if you are happy.
